zkevm-circuits/src/taiko_pi_circuit.rs

```diff
@@ -306,6 +669,431 @@ impl<F: Field> SubCircuitConfig<F> for TaikoPiCircuitConfig<F> {
+                let timestamp_len = get_len(TIMESTAMP_RLP_OFFSET + TIMESTAMP_SIZE);
+                let base_fee_len = get_len(BASE_FEE_RLP_OFFSET + BASE_FEE_SIZE);
+                // Only check the LSB of the length (the MSB is always 0x02!).
+                cb.require_equal(
+                    "total_len",
+                    meta.query_advice(blk_hdr_rlp, Rotation(TOTAL_LENGTH_OFFSET)),
+                    0x0F.expr()
+                        + number_len
+                        + gas_limit_len
+                        + gas_used_len
+                        + timestamp_len
+                        + base_fee_len,
+                );
+            });
+
+            // Leading zeros artificical headers are not part of the RLC calculation
+            let q_number_next = meta.query_fixed(q_number, Rotation::next());
+            let q_number_after_next = meta.query_fixed(q_number, Rotation(2));
+            let q_var_field_256_next = meta.query_fixed(q_var_field_256, Rotation::next());
+            let q_var_field_256_after_next = meta.query_fixed(q_var_field_256, Rotation(2));
+            let is_number_header =
+                and::expr([not::expr(q_number_next.expr()), q_number_after_next.expr()]);
+            let is_var_field_header = and::expr([
+                not::expr(q_var_field_256_next.expr()),
+                q_var_field_256_after_next.expr(),
+            ]);
+            let is_number_zero =
+                meta.query_advice(blk_hdr_is_leading_zero, Rotation((NUMBER_SIZE + 1) as i32));
+            let is_var_field_zero =
+                meta.query_advice(blk_hdr_is_leading_zero, Rotation((WORD_SIZE + 1) as i32));
+            let rlp_short_or_zero = rlp_is_short.is_lt(meta, Some(Rotation::next()));
+            // Artificial headers exist for header fields with short values greater than
+            // zero
+            let is_artificial_header = and::expr([
+                rlp_short_or_zero.expr(),
+                or::expr([
+                    and::expr([is_number_header, not::expr(is_number_zero.expr())]),
+                    and::expr([is_var_field_header, not::expr(is_var_field_zero.expr())]),
+                ]),
+            ]);
+            let no_rlc = or::expr([is_leading_zero_next.expr(), is_artificial_header]);
+
+            let do_rlc_val = select::expr(no_rlc, 0.expr(), 1.expr());
+            cb.condition(q_enabled.expr(), |cb| {
+                cb.require_equal(
+                    "skip leading zeros and artifical headers in RLC ",
+                    meta.query_advice(blk_hdr_do_rlc_acc, Rotation::cur()),
+                    do_rlc_val,
+                );
+            });
+
+            // Decode RLC and Hi/Lo field values
+            for (selector, values, r_mul) in [
+                            (q_hi_next.expr(), q_reconstruct_cur.expr(), 2_u64.pow(8).expr()),
+                            (q_lo_next.expr(), q_lo_cur, 2_u64.pow(8).expr()),
```

Comment:
`r_mul` is the same value twice so can just be used as is in the loop or no?

Reply:
in the line below, there is a third item in the iterator that has an `r_mul` value of `challenges.evm_word().expr()`

Reply:
Oh sorry, the somewhat strange formatting threw me off and was only looking at the code, not what it was doing.